### PR TITLE
Update UI reset functionality

### LINKED ISSUE #2399: Publish `AgentShutdownEvent`
## Description

The `Agent` models need to be updated with a `stop_time` when agents shut down.

## Goals/Epic/Context

Replace legacy `Monkey` model with new `Agent` model
Replace legacy `NodeService` and "node" models and endpoints with new `Machine`/`Agent`/`Node` models

## Tasks

- [x] Add an `AgentShutdownEvent` that has `stop_time` (0d) - @shreyamalviya 
- [x] Modify the agent to publish `AgentShutdownEvent` (Leave shutdown telem for now) (0d) - @shreyamalviya 
- [x] Subscribe to `AgentShutdownEvent` events in the Island and update the `Agent` models (0d) - @shreyamalviya 
- [ ] Update the UI's "clear data"/reset functionality to use "/api/agents" instead of "/api/root" (0.25d) - @shreyamalviya 
- [x] Add `AgentEventForwarder.flush()` (0d) - @mssalvatore

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes #2399 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [ ] ~Added relevant unit tests?~
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running the Island and Agent and making sure that everything works as expected
* [ ] ~If applicable, add screenshots or log transcripts of the feature working~

### CODE REVIEW

File: monkey/monkey_island/cc/setup/agent_event_handlers.py [hunk 1]
```diff
@@ -31,7 +35,9 @@ def _subscribe_and_store_to_event_repository(container: DIContainer):
         container.resolve(ICredentialsRepository)
     )
     agent_event_queue.subscribe_type(CredentialsStolenEvent, save_stolen_credentials_subscriber)
-    agent_event_queue.subscribe_type(AgentShutdownEvent, update_agent_shutdown_status)
+    agent_event_queue.subscribe_type(
+        AgentShutdownEvent, update_agent_shutdown_status(container.resolve(IAgentRepository))
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/setup/agent_event_handlers.py:39]: Maybe extract `container.resolve(IAgentRepository)` to a variable like it's done with `agent_event_queue`

File: monkey/monkey_island/cc/ui/src/components/pages/ReportPage.js [hunk 2]
```diff
@@ -95,18 +95,28 @@ class ReportPageComponent extends AuthComponent {
...
 
   updateMonkeysRunning = () => {
+    let any_agent_exists = false;
+    this.authFetch('/api/agents')
+      .then(res => res.json())
+      .then(res => {
+        if (res.length > 0) {
+          any_agent_exists = true;
+        }
+    });
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/pages/ReportPage.js:105]: This code is duplicated in `Main.tsx`. It should also use `IslandHttpClient`.

**Review by VakarisZ — COMMENTED** [review]
In `monkey/monkey_island/cc/ui/src/components/ui-components/IslandResetModal.tsx` it allows reset only if all agents stopped running. The way this is determined is by the old `Monkey.is_alive` model. The **new** way of determining if agents are dead is by comparing timestamps, right? Is this out of scope?

File: monkey/monkey_island/cc/ui/src/components/Main.tsx [hunk 3]
```diff
@@ -96,10 +96,23 @@ class AppComponent extends AuthComponent {
             if (this.state.islandMode === "unset") {
               return
             }
+
+            // check if any Agent was ever run
+            let any_agent_exists = false;
+            this.authFetch('/api/agents')
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/Main.tsx:102]: The fact that we need to download the whole collection just to check if any agents ran is a bit disturbing. If continue this route we'll soon need redux/global store

Comment (mssalvatore) [reply monkey/monkey_island/cc/ui/src/components/Main.tsx:102]: We discussed a bit over webex. This is ultimately a scalability concern and there are plans for addressing it.

**Review by mssalvatore — APPROVED** [review]
Approved once the [code duplication comment](https://github.com/guardicore/monkey/pull/2420/files#r993376685) is addressed.

File: monkey/monkey_island/cc/ui/src/components/utils/ServerUtils.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    return any_agent_exists;
+}
+
+export function didAllAgentsShutdown() {
+    let all_agents_shutdown = true;
+    let all_agents = _getAllAgents();
+    for (let idx in all_agents) {
+        let agent = all_agents[idx];
+        if (agent.stop_time === null) {
+            all_agents_shutdown = false;
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/utils/ServerUtils.js:18]: return false here, no need to keep looping

File: monkey/monkey_island/cc/services/node.py [hunk 5]
```diff
@@ -309,19 +308,6 @@ def set_node_exploited(node_id):
...
-        return any(not monkey.is_dead() for monkey in all_monkeys)
-
-    @staticmethod
-    def is_any_monkey_exists():
-        return mongo.db.monkey.find_one({}) is not None
-
-    @staticmethod
-    def is_monkey_finished_running():
-        return NodeService.is_any_monkey_exists() and not NodeService.is_any_monkey_alive()
-
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/node.py:324]: ❤️

File: monkey/monkey_island/cc/services/infection_lifecycle.py [hunk 6]
```diff
@@ -10,9 +10,8 @@
 logger = logging.getLogger(__name__)
 
 
-def get_completed_steps():
-    is_any_exists = NodeService.is_any_monkey_exists()
-    infection_done = NodeService.is_monkey_finished_running()
+def get_completed_steps(agent_repository: IAgentRepository):
+    infection_done = _is_infection_done(agent_repository)
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/infection_lifecycle.py:14]: If we do this then what's the point of changes in `Main.tsx`?

Comment (VakarisZ) [reply monkey/monkey_island/cc/services/infection_lifecycle.py:14]: I think `main.tsx` overrides `infection_done` step, I don't think we need to correct it in the back end

Comment (shreyamalviya) [reply monkey/monkey_island/cc/services/infection_lifecycle.py:14]: The infection lifecycle still uses it for report generation.

**Pull request merged**